pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,377 @@
+	errs := []error{}
+	result := reconcile.Result{}
+
+	logrus.Infof("reconciling request: %v", request)
+
+	// Get the current dns state.
+	dns := &operatorv1.DNS{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, dns); err != nil {
+		if errors.IsNotFound(err) {
+			// This means the dns was already deleted/finalized and there are
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
+			logrus.Infof("cluster dns not found; reconciliation will be skipped for request: %v", request)
+		} else {
+			errs = append(errs, fmt.Errorf("failed to get cluster dns %s: %v", request, err))
+		}
+		dns = nil
+	}
+
+	if dns != nil {
+		if dns.Name != DefaultDNSController {
+			// We do not want to return this error to avoid re-triggering the event
+			logrus.Errorf("skipping unexpected cluster dns %s", dns.Name)
+			return result, utilerrors.NewAggregate(errs)
+		}
+		// Ensure we have all the necessary scaffolding on which to place dns instances.
+		if err := r.ensureDNSNamespace(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to ensure dns namespace: %v", err))
+		}
+
+		if dns.DeletionTimestamp != nil {
+			// Handle deletion.
+			if err := r.ensureOpenshiftExternalNameServiceDeleted(); err != nil {
+				errs = append(errs, fmt.Errorf("failed to delete external name for openshift service: %v", err))
+			}
+			if err := r.ensureDNSDeleted(dns); err != nil {
+				errs = append(errs, fmt.Errorf("failed to ensure deletion for cluster dns %s: %v", dns.Name, err))
+			}
+
+			if len(errs) == 0 {
+				// Clean up the finalizer to allow the cluster dns to be deleted.
+				updated := dns.DeepCopy()
+				if slice.ContainsString(dns.Finalizers, DNSControllerFinalizer) {
+					updated.Finalizers = slice.RemoveString(updated.Finalizers, DNSControllerFinalizer)
+					if err := r.client.Update(context.TODO(), updated); err != nil {
+						errs = append(errs, fmt.Errorf("failed to remove finalizer from cluster dns %s: %v", dns.Name, err))
+					}
+				}
+			}
+		} else if err := r.enforceDNSFinalizer(dns); err != nil {
+			errs = append(errs, fmt.Errorf("failed to enforce finalizer for cluster dns %s: %v", dns.Name, err))
+		} else {
+			// Handle everything else.
+			if err := r.ensureClusterDNS(dns); err != nil {
+				errs = append(errs, fmt.Errorf("failed to ensure cluster dns %s: %v", dns.Name, err))
+			} else if err := r.ensureExternalNameForOpenshiftService(); err != nil {
+				errs = append(errs, fmt.Errorf("failed to ensure external name for openshift service: %v", err))
+			}
+		}
+	}
+
+	// TODO: Should this be another controller?
+	if err := r.syncOperatorStatus(); err != nil {
+		errs = append(errs, fmt.Errorf("failed to sync operator status: %v", err))
+	}
+
+	if len(errs) > 0 {
+		logrus.Errorf("failed to reconcile request %s: %v", request, utilerrors.NewAggregate(errs))
+	}
+	return result, utilerrors.NewAggregate(errs)
```

Comment:
Logging and returning `utilerrors.NewAggregate(errs)` seems redundant—are we doing that because the controller's logs get eaten? If so, please explain in a comment.

Reply:
Yes, controller's logs get eaten in this case. Added a comment for this.